concordium-consensus/src/Concordium/KonsensusV1/Types.hs

```diff
@@ -0,0 +1,867 @@
+blockSignatureMessageBytes = Hash.hashToByteString . blockHash
+
+-- |Verify that a block is correctly signed by the baker key present in the block.
+-- (This does not authenticate that the key corresponds to a genuine baker.)
+verifyBlockSignature :: (BakedBlockData b, HashableTo BlockHash b) => b -> Bool
+verifyBlockSignature b =
+    BlockSig.verify
+        (blockBakerKey b)
+        (blockSignatureMessageBytes $ getHash b)
+        (blockSignature b)
+
+-- |Sign a block hash as a baker.
+signBlockHash :: BakerSignPrivateKey -> BlockHash -> BlockSignature
+signBlockHash privKey bh = BlockSig.sign privKey (blockSignatureMessageBytes bh)
+
+-- |Sign a block as a baker.
+signBlock :: BakerSignPrivateKey -> BakedBlock -> SignedBlock
+signBlock privKey sbBlock = SignedBlock{..}
+  where
+    sbHash = getHash sbBlock
+    sbSignature = signBlockHash privKey sbHash
+
+-- |Message used to generate the block nonce with the VRF.
+blockNonceMessage :: LeadershipElectionNonce -> Round -> BS.ByteString
+blockNonceMessage leNonce rnd = runPut $ do
+    putByteString "NONCE"
+    put leNonce
+    put rnd
+
+-- |Generate the block nonce.
+computeBlockNonce :: LeadershipElectionNonce -> Round -> BakerElectionPrivateKey -> BlockNonce
+computeBlockNonce leNonce rnd key = VRF.prove key (blockNonceMessage leNonce rnd)
+
+-- |Verify a block nonce.
+verifyBlockNonce :: LeadershipElectionNonce -> Round -> BakerElectionVerifyKey -> BlockNonce -> Bool
+verifyBlockNonce leNonce rnd verifKey = VRF.verify verifKey (blockNonceMessage leNonce rnd)
+
+-- |The 'BlockHeader' consists of basic information about the block that is used in the generation
+-- of the block hash. This data is near the root of the Merkle tree constructing the block hash.
+data BlockHeader = BlockHeader
+    { -- |Round number of the block.
+      bhRound :: !Round,
+      -- |Epoch number of the block.
+      bhEpoch :: !Epoch,
+      -- |Hash of the parent block.
+      bhParent :: !BlockHash
+    }
+    deriving (Eq)
+
+-- |The block header for a 'BakedBlock'.
+bbBlockHeader :: BakedBlock -> BlockHeader
+bbBlockHeader BakedBlock{..} =
+    BlockHeader
+        { bhRound = bbRound,
+          bhEpoch = bbEpoch,
+          bhParent = qcBlock bbQuorumCertificate
+        }
+
+-- |The hash of a 'BlockHeader'. This is combined with the 'BlockQuasiHash' to produce a
+-- 'BlockHash'.
+newtype BlockHeaderHash = BlockHeaderHash {theBlockHeaderHash :: Hash.Hash}
+    deriving (Eq, Ord, Show, Serialize)
+
+instance HashableTo BlockHeaderHash BlockHeader where
+    getHash BlockHeader{..} = BlockHeaderHash $ Hash.hash $ runPut $ do
+        put bhRound
+        put bhEpoch
+        put bhParent
+
+instance HashableTo BlockHeaderHash BakedBlock where
+    getHash = getHash . bbBlockHeader
+
+-- |Hash of a block's contents. This is combined with the 'BlockHeaderHash' to produce a
+-- 'BlockHash'.
+newtype BlockQuasiHash = BlockQuasiHash {theBlockQuasiHash :: Hash.Hash}
```

Comment:
Perhaps it would be better to give this a name that you use in the description. Like `BlockContentsHash`.

`Quasi` does not convey any meaning.